### PR TITLE
fix(bits): prevent BitArray.UnmarshalJSON from crashing on 0 bits in the JSON

### PR DESCRIPTION
**Pull request by melekes**
This change fixes a bug in which BitArray.UnmarshalJSON hadn't accounted for the fact that invoking NewBitArray(<=0) returns nil and hence when dereferenced would crash with a runtime nil pointer dereference. This bug was found by my security analysis and fuzzing too.

Author: @odeke-em 

Fixes https://github.com/cometbft/cometbft/issues/2658

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] ~~Updated relevant documentation (`docs/` or `spec/`) and code comments~~
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Review by cason — APPROVED** [review]
Not clear what the use case for the introduced test unit, but it fixes it.

File: internal/bits/bit_array_test.go [hunk 1]
```diff
@@ -286,3 +286,18 @@ func TestBitArrayProtoBuf(t *testing.T) {
...
+func TestUnmarshalJSONDoesntCrashOnZeroBits(t *testing.T) {
+	type indexCorpus struct {
+		BitArray *BitArray `json:"ba"`
+		Index    int       `json:"i"`
+	}
+
+	ic := new(indexCorpus)
+	blob := []byte(`{"BA":""}`)
+	err := json.Unmarshal(blob, ic)
+	require.Nil(t, err, "the returned error must be nil")
```
Comment (sergio-mena) [inline internal/bits/bit_array_test.go:301]: Use `require.NoError` ?

**Review by sergio-mena — APPROVED** [review]
👍 I agree this can (should!) be backported

**Comment by faddat** [comment]
Fantastic

**Pull request merged**